src/dataflow/src/sink/kafka.rs

```diff
@@ -307,90 +511,34 @@ pub fn produce_to_kafka<G>(
-    // Sets the buffer size to be 16 GB (note that this setting is in KB)
-    config.set("queue.buffering.max.kbytes", &format!("{}", 16 << 20));
-
-    // Set the max messages buffered by the producer at any time to 10MM which
-    // is the maximum allowed value
-    config.set("queue.buffering.max.messages", &format!("{}", 10_000_000));
-
-    // Make the Kafka producer wait at least 10 ms before sending out MessageSets
-    // TODO(rkhaitan): experiment with different settings for this value to see
-    // if it makes a big difference
-    config.set("queue.buffering.max.ms", &format!("{}", 10));
-
-    for (k, v) in connector.config_options.iter() {
-        // We explicitly reject `statistics.interval.ms` here so that we don't
-        // flood the INFO log with statistics messages.
-        // TODO: properly support statistics on Kafka sinks
-        // We explicitly reject 'isolation.level' as it's a consumer property
-        // and, while benign, will fill the log with WARN messages
-        if k != "statistics.interval.ms" && k != "isolation.level" {
-            config.set(k, v);
-        }
-    }
-
-    let transactional = if connector.exactly_once {
-        // TODO(aljoscha): this only works for now, once there's an actual
-        // Kafka producer on each worker they would step on each others toes
-        let transactional_id = format!("mz-producer-{}", connector.topic);
-        config.set("transactional.id", transactional_id);
-        true
-    } else {
-        false
-    };
-
-    let shutdown_flag = Arc::new(AtomicBool::new(false));
     let mut builder = OperatorBuilder::new(name.clone(), stream.scope());
+    let activator = stream
+        .scope()
+        .activator_for(&builder.operator_info().address[..]);
 
-    let s = {
-        let metrics = SinkMetrics::new(
-            &connector.topic,
-            &id.to_string(),
-            &stream.scope().index().to_string(),
-        );
-
-        let producer = config
-            .create_with_context::<_, ThreadedProducer<_>>(SinkProducerContext::new(
-                metrics.clone(),
-                shutdown_flag.clone(),
-            ))
-            .expect("creating kafka producer for kafka sinks failed");
+    let shutdown_flag = Arc::new(AtomicBool::new(false));
 
-        let activator = stream
-            .scope()
-            .activator_for(&builder.operator_info().address[..]);
+    let mut s = KafkaSinkState::new(
+        connector,
+        name,
+        &id,
+        stream.scope().index().to_string(),
+        Arc::clone(&shutdown_flag),
+        activator,
+        write_frontier,
+    );
 
-        KafkaSink {
-            name,
-            shutdown_flag: shutdown_flag.clone(),
-            metrics,
-            producer,
-            activator,
-            txn_timeout: Duration::from_secs(5),
-        }
-    };
+    // Keep track of whether this operator/worker ever received updates. We
```

Comment:
this really feels like we want to pass a boolean that says `active_worker` or something like that at sink render time to indicate which operator should be the writer instead of making it implicit with the hashing

Reply:
I did it like this because it's less fragile than relying on knowing how timely hashes and exchanges data to workers. But I'm happy to change it since we already have that dependency for the write frontier stuff.

Reply:
> > My only question is: is emitting END records after every frontier update acceptable from a perf pov?
> 
> I wrote this one before:
> 
> > I didn't introduce any knobs for controlling when/how we write out END records. This can already be controlled via the timestamping frequency (for RT sources), and I didn't want to introduce any additional settings.
> 
> I'm also not sure about it, though. With BYO timestamping, you get one `END` record in the output per `END` record in the input. With RT timestamping you get an `END` record per minted binding, which is controllable by the user. (and defaults to one per second)

Oh right sorry! yeah I think that reasoning makes sense